### ISSUE TITLE
Introduce FunctionRef equivalent for cluster sharding

### ISSUE THREAD

**Issue by Horusiath** [akka/akka#25030]
The idea here is to add an abstraction over entities managed by cluster sharding feature, that would work over cluster sharding in the same way, `FunctionRef` works currently over `ActorRef`. The further use case here would be to make a location transparent - as imagined by cluster sharding routing transparency - stream refs equivalent.

Of course this would come with some issues - as sharded entities are potentially eternal, an adaptation to things like stream completion would be necessary (maybe using the same rule as in case of deleted keys in DistributedData extension).

**Comment by patriknw** [comment]
I'm not sure I follow. In the Typed APIs we added [EntityRef](https://doc.akka.io/docs/akka/current/typed/cluster-sharding.html) and perhaps that is what you are looking for?

**Comment by Horusiath** [comment]
Yes, it's related to `EntityRef`s. However they live on a caller side and are not present (?) in core untyped akka, which doesn't have them but also requires Actor representation to work with. This proposal aims to span sharding capabilities not only over actors, but any kind of functionality that can be represented as a function (desirably to make things like sharded streams possible).

**Comment by Horusiath** [comment]
@svozniuk stream refs only allows to maintain location transparency. This is still far from sharding/singletons. If a machine on which stream was materialized, dies for any reason, you need to manually acknowledge that fact and move graph to another one. Same thing concerns rebalancing. Here the idea was to expose potential streaming capability on several machines and let them to switch from one to another in case of failure.
